src/proxy_lib/proxy_lib.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+#include <umf/pools/pool_jemalloc.h>
+#include <umf/providers/provider_os_memory.h>
+
+#include "proxy_lib.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+
+static UTIL_ONCE_FLAG alloc_hooks_are_initialized = UTIL_ONCE_FLAG_INIT;
+
+static umf_memory_provider_handle_t OS_memory_provider = NULL;
+static umf_memory_pool_handle_t Proxy_pool = NULL;
+
+// it protects us from recursion in umfPool*()
+static __TLS int was_called_from_umfPool = 0;
+
+malloc_func_t System_malloc_func;
```

Comment:
Why do we use original malloc functions instead of just using base_alloc or linear alloc? WIth one of those, we wouldn't even need to load malloc* symbols, right?

Reply:
See the comment added just now:
```c
/*
 * The UMF proxy library uses three memory allocators:
 * 1) the internal linear base allocator (LBA) based on the anonymous mapped memory
 * 2) the system allocator (libc malloc on Linux for example)
 * 3) the main one - UMF pool allocator.
 *
 * Ad 1)
 * The internal linear base allocator (LBA) is used from the very beginning
 * to the start of the constructor of the proxy library to allocate memory
 * for OS specific structures used during loading and starting applications
 * (for example _dl_init() and _dl_fini() on Linux storing arrays of all
 * constructors and destructors that have to be called). That memory will be leaked
 * on purpose (we believe OS will free it), because we cannot free the memory
 * containing array of destructors that have to be called at the end.
 * There are only 4 such allocations on Linux now:
 * - realloc(0x0, 480)
 * - malloc(20)
 * - malloc(24)
 * - malloc(72704)
 *
 * Ad 2)
 * The system allocator is used from the start of the constructor of the proxy library
 * to the creation of a UMF pool to allocate memory needed by umfMemoryProviderCreate()
 * and umfPoolCreate(). It will be freed by the system allocator.
 *
 * Ad 3)
 * The UMF pool allocator (the main one) is used from the creation to the destruction
 * of a UMF pool to allocate memory needed by an application. It should be freed
 * by an application.
 */
```

Reply:
> See the comment added just now:
> 
> ```c
> /*
>  * The UMF proxy library uses three memory allocators:
>  * 1) the internal linear base allocator (LBA) based on the anonymous mapped memory
>  * 2) the system allocator (libc malloc on Linux for example)
>  * 3) the main one - UMF pool allocator.
>  *
>  * Ad 1)
>  * The internal linear base allocator (LBA) is used from the very beginning
>  * to the start of the constructor of the proxy library to allocate memory
>  * for OS specific structures used during loading and starting applications
>  * (for example _dl_init() and _dl_fini() on Linux storing arrays of all
>  * constructors and destructors that have to be called). That memory will be leaked
>  * on purpose (we believe OS will free it), because we cannot free the memory
>  * containing array of destructors that have to be called at the end.
>  * There are only 4 such allocations on Linux now:
>  * - realloc(0x0, 480)
>  * - malloc(20)
>  * - malloc(24)
>  * - malloc(72704)
>  *
>  * Ad 2)
>  * The system allocator is used from the start of the constructor of the proxy library
>  * to the creation of a UMF pool to allocate memory needed by umfMemoryProviderCreate()
>  * and umfPoolCreate(). It will be freed by the system allocator.
>  *
>  * Ad 3)
>  * The UMF pool allocator (the main one) is used from the creation to the destruction
>  * of a UMF pool to allocate memory needed by an application. It should be freed
>  * by an application.
>  */
> ```
> 
> So the short answer is: to not leak too much memory.

I still don't quite understand why we would leak memory when using base_alloc. Are you worried that there are multiple different allocation sizes and creating a single base_alloc pool would waste too much memory? If so, we can probably just use the version from here: https://github.com/oneapi-src/unified-memory-framework/pull/203 right?

Reply:
@igchor On the other hand - what is wrong in using original malloc functions and loading malloc* symbols?

Reply:
> I still don't quite understand why we would leak memory when using base_alloc. Are you worried that there are multiple different allocation sizes and creating a single base_alloc pool would waste too much memory? If so, we can probably just use the version from here: #203 right? (we might need to just tweak the allocation class sizes)

We cannot use the version from #203 since `free()` requires the size of the allocation in this version, but we do not know that (allocations are done inside hwloc for example).
The only solution I see now is to use the second instance of the linear base allocator instead of the system one.

Reply:
@igchor I have replaced the system allocator with the second instance of the linear base allocator - please review now.

Reply:
Thanks, you are right about #203, we would need to pass size. 

> @igchor On the other hand - what is wrong in using original malloc functions and loading malloc* symbols?

I think loading malloc symbols just adds complexity (especially since the code is different for Windows) and we already have allocators (linear_alloc or even other memory pools that we could use with a simplified memory provider: one that doesn't need to allocate anything and just calls mmap/VirtualReserve).

Reply:
@igchor is it resolved now?